resolve.go

```diff
@@ -164,44 +167,35 @@ func resolveServer(serverName ServerName, needWellKnown bool) (results []Resolut
 	if port != -1 {
 		results = []ResolutionResult{
 			ResolutionResult{
-				Destination: string(serverName),
-				Host:        serverName,
-				Name:        host,
+				Destination:   string(serverName),
+				Host:          serverName,
+				TLSServerName: host,
 			},
 		}
 
 		return
 	}
 
-	if needWellKnown {
+	if checkWellKnown {
 		// 3. If the hostname is not an IP literal
 		var result *WellKnownResult
 		result, err = LookupWellKnown(serverName)
 		if err == nil {
-			if len(result.NewAddress) > 0 {
-				// We don't want to check .well-known on the result
-				return resolveServer(result.NewAddress, false)
-			}
+			// We don't want to check .well-known on the result
+			return resolveServer(result.NewAddress, false)
 		}
 	}
 
-	// handleNoWellKnown implements steps 4 and 5 of the algorithm (as well as
-	// 3.3 and 3.4)
-	return handleNoWellKnown(serverName, host)
+	return handleNoWellKnown(serverName, host), nil
 }
 
-func handleNoWellKnown(serverName ServerName, host string) (results []ResolutionResult, err error) {
-	dnsResults, err := LookupServer(serverName)
```

Comment:
can we get rid of LookupServer now?

Reply:
Nope, the federation tester uses it to gather DNS results.

Reply:
the federation tester needs fixing, then. We shouldn't have stuff in gomatrixserverlib which is only useful for the fed tester. (In any case, it doesn't look like the right thing for the federation tester, either.)

However, I'm happy for this to be left for a separate PR. A comment on it saying that it shouldn't be used would be useful though.